indexing-service/src/main/java/io/druid/indexing/common/task/SinglePhaseParallelIndexTaskRunner.java

```diff
@@ -0,0 +1,624 @@
+      completeSubTaskSpecs.forEach(spec -> subTaskSpecMap.put(spec.getId(), spec));
+      return Response.ok(new ArrayList<>(subTaskSpecMap.values())).build();
+    } else {
+      return Response.ok(Collections.emptyList()).build();
+    }
+  }
+
+  @GET
+  @Path("/subtaskspecs/running")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getRunningSubTaskSpecs(@Context final HttpServletRequest req)
+  {
+    IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
+    final List<SubTaskSpec<ParallelIndexSubTask>> runningSubTaskSpecs = taskMonitor == null ?
+                                                                        Collections.emptyList() :
+                                                                        taskMonitor.getRunningSubTaskSpecs();
+    return Response.ok(runningSubTaskSpecs).build();
+  }
+
+  @GET
+  @Path("/subtaskspecs/complete")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getCompleteSubTaskSpecs(@Context final HttpServletRequest req)
+  {
+    IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
+    final List<SubTaskSpec<ParallelIndexSubTask>> completeSubTaskSpecs =
+        taskMonitor == null ?
+        Collections.emptyList() :
+        taskMonitor.getCompleteSubTaskSpecs();
+    return Response.ok(completeSubTaskSpecs).build();
+  }
+
+  @GET
+  @Path("/subtaskspec/{id}")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSubTaskSpec(@PathParam("id") String id, @Context final HttpServletRequest req)
+  {
+    IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
+    if (taskMonitor != null) {
+      // Running tasks should be checked first because, in taskMonitor, subTaskSpecs are removed from runningTasks after
+      // adding them to taskHistory.
+      final MonitorEntry monitorEntry = taskMonitor.getRunningTaskMonitorEntory(id);
+      final TaskHistory<ParallelIndexSubTask> taskHistory = taskMonitor.getCompleteSubTaskSpecHistory(id);
+      final SubTaskSpec<ParallelIndexSubTask> subTaskSpec;
+
+      if (monitorEntry != null) {
+        subTaskSpec = monitorEntry.getSpec();
+      } else {
+        if (taskHistory != null) {
+          subTaskSpec = taskHistory.getSpec();
+        } else {
+          subTaskSpec = null;
+        }
+      }
+
+      if (subTaskSpec == null) {
+        return Response.status(Response.Status.NOT_FOUND).build();
+      } else {
+        return Response.ok(subTaskSpec).build();
+      }
+    } else {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+  }
+
+  @GET
+  @Path("/subtaskspec/{id}/state")
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response getSubTaskState(@PathParam("id") String id, @Context final HttpServletRequest req)
+  {
+    IndexTaskUtils.datasourceAuthorizationCheck(req, Action.READ, getDataSource(), authorizerMapper);
```

Comment:
these should end up returning `403` or similar right, not throwing and ending up in a `500`?

Reply:
I think you're correct.